src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -1,149 +1,203 @@
 
-  override def getValue(rvb: RegionValueBuilder): Unit
-}
+    alleles = new Array[String](nAlleles)
 
-class BgenRecordV12(compressed: Boolean, nSamples: Int,
-  includeGT: Boolean, includeGP: Boolean, includeDosage: Boolean) extends BgenRecord {
-  var expectedDataSize: Int = _
-  var expectedNumAlleles: Int = _
+    val ref = bfis.readLengthAndString(4)
+    alleles(0) = ref
 
-  def setExpectedDataSize(size: Int) {
-    this.expectedDataSize = size
-  }
+    var aIdx = 1
+    while (aIdx < nAlleles) {
+      alleles(aIdx) = bfis.readLengthAndString(4)
+      aIdx += 1
+    }
 
-  def setExpectedNumAlleles(n: Int) {
-    this.expectedNumAlleles = n
-  }
+    val recodedContig = contig match {
+      case "23" => "X"
+      case "24" => "Y"
+      case "25" => "X"
+      case "26" => "MT"
+      case x => x
+    }
 
-  override def getValue(rvb: RegionValueBuilder) {
-    require(input != null, "called getValue before serialized value was set")
+    dataSize = bfis.readInt()
 
-    val a = if (compressed) decompress(input, expectedDataSize) else input
-    val reader = new ByteArrayReader(a)
+    if (includeAnyEntryFields) {
+      data = if (compressed) {
+        val expectedDataSize = bfis.readInt()
+        val input = bfis.readBytes(dataSize - 4)
+        decompress(input, expectedDataSize)
+      } else {
+        bfis.readBytes(dataSize)
+      }
+      val reader = new ByteArrayReader(data)
+
+      val nRow = reader.readInt()
+      if (nRow != nSamples)
+        fatal("row nSamples is not equal to header nSamples $nRow, $nSamples")
+
+      val nAlleles2 = reader.readShort()
+      if (nAlleles == nAlleles2)
+        fatal(s"""Value for `nAlleles' in genotype probability data storage is
+                 |not equal to value in variant identifying data. Expected
+                 |$nAlleles but found $nAlleles2 at $lid.""".stripMargin)
+
+      val minPloidy = reader.read()
+      val maxPloidy = reader.read()
+
+      if (minPloidy != 2 || maxPloidy != 2)
+        fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+      var i = 0
+      while (i < nSamples) {
+        val ploidy = reader.read()
+        if ((ploidy & 0x3f) != 2)
+          fatal(s"Ploidy value must equal to 2. Found $ploidy.")
+        i += 1
+      }
 
-    val nRow = reader.readInt()
-    if (nRow != nSamples)
-      fatal("row nSamples is not equal to header nSamples $nRow, $nSamples")
+      val phase = reader.read()
+      if (phase != 0 && phase != 1)
+        fatal(s"Value for phase must be 0 or 1. Found $phase.")
+      val isPhased = phase == 1
 
-    val nAlleles = reader.readShort()
-    if (nAlleles != expectedNumAlleles)
-      fatal(s"""Value for `nAlleles' in genotype probability data storage is
-               |not equal to value in variant identifying data. Expected
-               |$expectedNumAlleles but found $nAlleles.""".stripMargin)
-    if (nAlleles != 2)
```

Comment:
pulled this check up to the earlier read of `nAlleles` (which is stored twice for some reason??)